File: runners/python/steps/client_steps.py
from behave import given, when, then
from openfga_sdk import ClientConfiguration
from openfga_sdk.credentials import Credentials, CredentialConfiguration
import asyncio


@given('I have a client configured with store "{store_id}"')
def step_client_configured_with_store(context, store_id):
    config = {'store_id': store_id}
    asyncio.run(context.test_context.create_client(config))


@given('I have a client configured with')
def step_client_configured_with_table(context):
    config = {}
    
    for row in context.table:
        key = list(row.headings)[0]
        value = row[key]
        
        if key.lower() == 'apiurl':
            # ApiUrl is handled in create_client
            pass
        elif key.lower() == 'storeid':
            config['store_id'] = value
        elif key.lower() == 'authorizationmodelid':
            config['authorization_model_id'] = value
    
    asyncio.run(context.test_context.create_client(config))


@given('I configure authentication with bearer token "{token}"')
def step_configure_bearer_token(context, token):
    context.test_context.configure_authentication('bearer', {'token': token})


@given('I configure client credentials authentication')
def step_configure_client_credentials(context):
    credentials = {}
    
    for row in context.table:
        key = list(row.headings)[0]
        value = row[key]
        credentials[key.lower()] = value
    
    context.test_context.configure_authentication('client_credentials', {
        'client_id': credentials.get('clientid', ''),
        'client_secret': credentials.get('clientsecret', ''),
        'api_token_issuer': credentials.get('apitokenissuer', ''),
        'api_audience': credentials.get('apiaudience', '')
    })


@given('I configure default headers')
def step_configure_default_headers(context):
    for row in context.table:
        key = list(row.headings)[0]
        value = row[key]
        context.test_context.headers[key] = value
    
    # Apply headers to existing client if available
    if context.test_context.client:
        # Note: OpenFGA Python SDK header setting would go here
        # Implementation depends on SDK capabilities
        pass


@given('I set the request header "{header}" to "{value}"')
def step_set_request_header(context, header, value):
    context.test_context.set_request_header(header, value)


@given('I configure retry settings')
def step_configure_retry_settings(context):
    retry_config = {}
    
    for row in context.table:
        key = list(row.headings)[0]
        value = row[key]
        
        if key.lower() == 'maxretries':
            retry_config['max_retries'] = int(value)
        elif key.lower() == 'backofftype':
            retry_config['backoff_type'] = value
        elif key.lower() == 'basedelay':
            retry_config['base_delay'] = int(value)
        elif key.lower() == 'circuitbreakerenabled':
            retry_config['circuit_breaker_enabled'] = value.lower() == 'true'
        elif key.lower() == 'circuitbreakerthreshold':
            retry_config['circuit_breaker_threshold'] = int(value)
        elif key.lower() == 'jitterenabled':
            retry_config['jitter_enabled'] = value.lower() == 'true'
    
    # Validate retry configuration
    if retry_config.get('max_retries', 0) < 0:
        raise ValueError('maxRetries must be non-negative')
    
    context.test_context.retry_config = retry_config
    context.test_context.retry_attempts = []
    context.test_context.retry_delays = []


@then('the client should be configured successfully')
def step_client_configured_successfully(context):
    assert context.test_context.client is not None, "Client is not configured"
    assert context.test_context.config is not None, "Client configuration is not set"

File: runners/js/features/step_definitions/client_steps.ts
import { Given, When, Then } from '@cucumber/cucumber';
import { TestWorld } from './world';
import { expect } from 'chai';
import { ClientConfiguration } from '@openfga/sdk';

// Client configuration steps
Given('I have a client configured with store {string}', async function(this: TestWorld, storeId: string) {
  await this.createClient({ storeId } as unknown as ClientConfiguration);
});

Given('I have a client configured with:', async function(this: TestWorld, dataTable: any) {
  const config: Partial<ClientConfiguration> = {};
  
  for (const row of dataTable.hashes()) {
    const key = Object.keys(row)[0];
    const value = Object.values(row)[0];
    
    switch (key) {
      case 'apiUrl':
        config.apiUrl = value as string;
        break;
      case 'storeId':
        config.storeId = value as string;
        break;
      case 'authorizationModelId':
        config.authorizationModelId = value as string;
        break;
    }
  }
  
  await this.createClient(config as ClientConfiguration);
});

Given('I configure authentication with bearer token {string}', async function(this: TestWorld, token: string) {
  this.configureAuthentication('bearer', { token });
});

Given('I configure client credentials authentication:', async function(this: TestWorld, dataTable: any) {
  const credentials: { [key: string]: string } = {};
  for (const row of dataTable.hashes()) {
    const key = Object.keys(row)[0];
    const value = Object.values(row)[0];
    credentials[key] = value as string;
  }
  
  this.configureAuthentication('client_credentials', {
    clientId: credentials.clientId,
    clientSecret: credentials.clientSecret,
    apiTokenIssuer: credentials.apiTokenIssuer,
    apiAudience: credentials.apiAudience
  });
});

Given('I configure default headers:', async function(this: TestWorld, dataTable: any) {
  for (const row of dataTable.hashes()) {
    const key = Object.keys(row)[0];
    const value = Object.values(row)[0];
    this.headers.set(key, value as string);
  }
  
  // Configure default headers using the TestWorld method
  await this.configureDefaultHeaders(this.headers);
});

Given('I set the request header {string} to {string}', function(this: TestWorld, key: string, value: string) {
  this.setRequestHeader(key, value);
});

// Retry configuration
Given('I configure retry settings:', function(this: TestWorld, dataTable: any) {
  const retryConfig: { [key: string]: any } = {};
  
  for (const row of dataTable.hashes()) {
    const key = Object.keys(row)[0];
    const value = Object.values(row)[0];
    
    switch (key) {
      case 'maxRetries':
        retryConfig.maxRetries = parseInt(value as string);
        break;
      case 'backoffType':
        retryConfig.backoffType = value;
        break;
      case 'baseDelay':
        retryConfig.initialDelay = parseInt(value as string);
        break;
      case 'circuitBreakerEnabled':
        retryConfig.circuitBreakerEnabled = value === 'true';
        break;
      case 'circuitBreakerThreshold':
        retryConfig.circuitBreakerThreshold = parseInt(value as string);
        break;
      case 'jitterEnabled':
        retryConfig.jitterEnabled = value === 'true';
        break;
    }
  }
  
  // Validate retry configuration
  if (retryConfig.maxRetries < 0) {
    throw new Error('maxRetries must be non-negative');
  }
  
  this.retryConfig = retryConfig;
  this.retryAttempts = 0;
  this.retryDelays = [];
});

Then('the client should be configured successfully', function() {
  expect(this.client).to.not.be.null;
  expect(this.config).to.not.be.null;
});
